Fix for mdtag issues with insertions

arahuja opened the pull request:
Sorry all, looks the code in #727 contains a few errors when it came to parsing mismatches mixed with insertions.  I added a quite a few tests to cover this case and other complex mdtags.

If you think there are interesting cases not covered in the new tests now let me know!

Thanks!

ryan-williams commented on adam-core/src/main/scala/org/bdgenomics/adam/util/MdTag.scala:
```diff
@@ -122,16 +135,16 @@ object MdTag {
+            cigarOperatorIndex = 0
           }
           case _ => {
             cigarIdx += 1
```
is the `cigarOperatorIndex = 0` not necessary here?

mind commenting these last two cases a little bit? I assume the main interesting cigar operator being passed over here is `I`/insertion, which MDTags don't care about... any other cases being handled here worth noting?

arahuja replied:
`cigarOperatorIndex = 0` is not necessary, it shouldn't be updated, but just added it for clarity.

The last case is insertion or clipped bases

ryan-williams replied:
So it doesn't matter whether you set `cigarOperatorIndex = 0` here or not? That's also confusing to me, I don't fully have my head around this.

Is the second-to-last case one that you specifically added for `N`s? Is that the main focus there?

Could you add comments clarifying which operators you expect to hit each of these last two? At this point in the `match` I'm struggling to keep in my head what's left.

arahuja replied:
Yes the second to last case is for `N`

ryan-williams commented on adam-core/src/test/scala/org/bdgenomics/adam/util/MdTagSuite.scala:
```diff
@@ -225,6 +225,131 @@ class MdTagSuite extends FunSuite {
+    assert(tag.toString === "2A4A2")
+  }
+
+  test("Get correct matches for mdtag with intron and deletion between mismatches") {
```
hm, copy mcpastington? same test as above, different name?

arahuja replied:
Yes! I should was trying to think of more cases, but didn't add the test actually

ryan-williams commented on adam-core/src/test/scala/org/bdgenomics/adam/util/MdTagSuite.scala:
```diff
@@ -225,6 +225,131 @@ class MdTagSuite extends FunSuite {
+    (9 to 10).foreach(locus => assert(tag.isMatch(locus)))
+
+    assert(tag.mismatches(11L) === 'A')
+
```
nit: missing a last "match" check on idx 12?

arahuja replied:
Yea figured it was repetitive at this point, but I'll add it

ryan-williams commented:
added some of my nits in https://github.com/hammerlab/adam/pull/2 including the check that the number of deleted bases is consistent between cigar and MDTag.

thoughts on some test cases verifying that IAEs are thrown when the various assumptions are broken?

arahuja commented:
@ryan-williams I fixed the dupe test and added a test case for the exception you added (mdtag and cigar disagreeing on deleted bases)

I also made the match statement explicit for the operators they were catching

re: `cigarOperatorIndex

ryan-williams commented:
cool, thanks @arahuja. LGTM once tests pass. Last complete build failed due to needing `./scripts/format-source` btw

arahuja commented:
Ah ok, good to know, I didn't realize this wasn't auto-formatting, I'll run that, squash etc.

fnothaft commented:
LGTM as well.
